### DISCUSSION TITLE
Configuring Keycloak

### DISCUSSION THREAD

**Discussion by CarlosIVI** [eclipse-theia/theia-cloud#217] [Q&A] [answered]
Hello,

I'm currently trying to run TheIA-Cloud on K8S on premise, I finally were able to make it work, but after trying to enable Keycloak option it starts to fail.

In the docs is not clear what is the minimal configuration expected before modifying the helm chart.

I would like to know which pre-steps are required before enabling Keycloak

Should I install it by my own?
Do I need another CNAME DN for Keycloak?
What is the role of the use of Keycloak js mentioned on keycloak.md

Thanks in advance

**Comment by jfaltermeier** [comment, marked as answer]
Hi, 

did you have a look at our Terraform charts? https://github.com/eclipsesource/theia-cloud/blob/main/terraform/terraform.md
With these charts we automate the whole infrastructure setup, including installing and configuring all dependencies, in a reproducible way. 
I would suggest to have a look at Terraform because it really makes maintaining infrastructure a lot easier, especially when you have to change existing infrastructure later on. 

I think based on the our existing terraform charts, you should be able to create a setup for your on premise cluster. 
In the `terraform/modules` directory we have reusable components. 
The `cluster_creation` subdirectory has charts to set up a Minikube or GKE cluster, however since you have a cluster already, this can be skipped. 

The `helm` subdirectory contains all installations done via helm: https://github.com/eclipsesource/theia-cloud/blob/main/terraform/modules/helm/main.tf
So basically it shows the values and chart information for the cert-manager, the nginx-ingress-controller, keycloak (https://github.com/eclipsesource/theia-cloud/blob/f6ce7317643f2121cfbc430b93d913e5d0ad8bf2/terraform/modules/helm/main.tf#L159) and the theia cloud installations. 

The `keycloak` subdirectory contains the information on how to set up the keycloak realm: https://github.com/eclipsesource/theia-cloud/blob/main/terraform/modules/keycloak/main.tf
It also adds two sample users. 

Regarding the Prerequisites:
You need a running Keycloak with a configured Realm. These information are then provided to the Theia Cloud Helm Charts via: https://github.com/eclipsesource/theia-cloud-helm/blob/ff0b6d87bd7cfba3589543f63b1df9a42783c390/charts/theia.cloud/values.yaml#L115
(Or in the Terraform charts: https://github.com/eclipsesource/theia-cloud/blob/f6ce7317643f2121cfbc430b93d913e5d0ad8bf2/terraform/modules/helm/main.tf#L233 and https://github.com/eclipsesource/theia-cloud/blob/f6ce7317643f2121cfbc430b93d913e5d0ad8bf2/terraform/modules/helm/theia-cloud.yaml#L26)

Regarding keycloak.js. We use this in our example landing-pages (e.g. https://github.com/eclipsesource/theia-cloud/blob/f6ce7317643f2121cfbc430b93d913e5d0ad8bf2/node/try-now-page/src/App.tsx#L80) to authenticate a user and make authenticated calls to the Rest API.

**Reply by CarlosIVI** [reply]
Thank you so much, I'll provide some update here once I can enable Keycloak on premise
